File: Chimera.cpp
/*
 * -----------------------------------------------------------------------------
 * Filename:      Chimera.cpp
 *
 * Author:        Qinzhong Tian
 *
 * Email:         tianqinzhong@qq.com
 *
 * Created Date:  2024-07-09
 *
 * Last Modified: 2024-11-18
 *
 * Description:
 *  This is the main entry for Chimera
 *
 * Version:
 *  1.3
 * -----------------------------------------------------------------------------
 */
#include <CLI11.hpp>
#include <ChimeraBuild.hpp>
#include <ChimeraClassify.hpp>
#include <buildConfig.hpp>
#include <classifyConfig.hpp>
#include <fstream>
#include <iostream>
#include <limits>
#include <stdexcept>
#include <string>
#include <algorithm>
#include <cctype>
#include <thread>

#if defined(__unix__) || defined(__APPLE__) || defined(__linux__)
#include <sys/utsname.h>
#endif

#ifdef CHIMERA_VERSION
#define VERSION_INFO CHIMERA_VERSION
#else
#define VERSION_INFO "unknown"
#endif

namespace {

std::string detect_os() {
#if defined(_WIN32)
  return "Windows";
#elif defined(__APPLE__)
  struct utsname info {};
  if (uname(&info) == 0) {
    return std::string(info.sysname) + " " + info.release;
  }
  return "macOS";
#elif defined(__linux__)
  std::ifstream osrelease("/etc/os-release");
  if (osrelease) {
    std::string line;
    while (std::getline(osrelease, line)) {
      constexpr char key[] = "PRETTY_NAME=";
      if (line.rfind(key, 0) == 0) {
        std::string value = line.substr(sizeof(key) - 1);
        if (!value.empty() && value.front() == '"' && value.back() == '"') {
          value = value.substr(1, value.size() - 2);
        }
        if (!value.empty()) {
          return value;
        }
      }
    }
  }
  struct utsname info {};
  if (uname(&info) == 0) {
    return std::string(info.sysname) + " " + info.release;
  }
  return "Linux";
#elif defined(__FreeBSD__)
  struct utsname info {};
  if (uname(&info) == 0) {
    return std::string(info.sysname) + " " + info.release;
  }
  return "FreeBSD";
#else
  return "Unknown";
#endif
}

std::string normalize_auto_min_length_help() {
  return "Minimum length sequence for building (auto => strobemer minimum span)";
}

uint16_t default_cli_threads() {
  unsigned int hardware_threads = std::thread::hardware_concurrency();
  if (hardware_threads == 0) {
    hardware_threads = 1;
  }
  const auto max_threads =
      static_cast<unsigned int>(std::numeric_limits<uint16_t>::max());
  if (hardware_threads > max_threads) {
    hardware_threads = max_threads;
  }
  return static_cast<uint16_t>(std::min<unsigned int>(hardware_threads, 192u));
}

std::string validate_min_length_option(std::string &input) {
  std::string lowered = input;
  std::transform(lowered.begin(), lowered.end(), lowered.begin(),
                 [](unsigned char ch) {
                   return static_cast<char>(std::tolower(ch));
                 });
  if (lowered == "auto") {
    input = "0";
    return {};
  }
  try {
    long long value = std::stoll(input);
    if (value < 0) {
      return "Minimum length must be >= 0 or 'auto'";
    }
  } catch (const std::exception &) {
    return "Minimum length must be an integer or 'auto'";
  }
  return {};
}

void normalize_auto_string(std::string &value) {
  std::transform(value.begin(), value.end(), value.begin(),
                 [](unsigned char ch) {
                   return static_cast<char>(std::tolower(ch));
                 });
  if (value.empty()) {
    value = "auto";
  }
}

void validate_build_config(ChimeraBuild::BuildConfig &buildConfig,
                           bool buildQuietRequested) {
  normalize_auto_string(buildConfig.taxonomy_kind);
  buildConfig.min_length =
      std::max<uint64_t>(buildConfig.min_length, buildConfig.strobemer_k);
  if (buildConfig.strobemer_w_min == 0 || buildConfig.strobemer_w_max == 0) {
    throw CLI::ValidationError("--strobe-w-min/w-max must be greater than 0");
  }
  if (buildConfig.strobemer_w_min > buildConfig.strobemer_w_max) {
    throw CLI::ValidationError("--strobe-w-min must be <= --strobe-w-max");
  }
  if (buildConfig.strobemer_k < 8) {
    throw CLI::ValidationError("--strobe-k must be >= 8");
  }
  if (buildConfig.strobemer_order != 2) {
    throw CLI::ValidationError("--strobe-order currently only supports value 2");
  }
  buildConfig.verbose = !buildQuietRequested;
}

} // namespace

int main(int argc, char **argv) {
  // Create the main application object
  CLI::App app{"Chimera - A versatile tool for metagenomic classification"};
  ChimeraBuild::BuildConfig buildConfig;
  ChimeraClassify::ClassifyConfig classifyConfig;

  bool show_version = false;
  app.add_flag("-v,--version", show_version, "Show version information");
  // Create subcommands
  auto build = app.add_subcommand("build", "Build a sequence database");
  auto classify = app.add_subcommand("classify", "Classify sequences");

  const uint16_t default_threads = default_cli_threads();

  bool buildQuietRequested = false;

  // Build
  build
      ->add_option("-i,--input", buildConfig.input_file,
                   "Input file for building")
      ->required()
      ->check(CLI::ExistingFile);
  build
      ->add_option("-o,--output", buildConfig.output_file,
                   "Output file for building")
      ->default_val("ChimeraDB");
  build
      ->add_option("--strobe-k", buildConfig.strobemer_k,
                   "Strobemer k-mer length")
      ->default_val(28);
  build
      ->add_option("--strobe-order", buildConfig.strobemer_order,
                   "Strobemer order (currently only 2 is supported)")
      ->default_val(2);
  build
      ->add_option("--strobe-w-min", buildConfig.strobemer_w_min,
                   "Strobemer minimum window")
      ->default_val(12);
  build
      ->add_option("--strobe-w-max", buildConfig.strobemer_w_max,
                   "Strobemer maximum window")
      ->default_val(32);
  auto *min_length_option = build
      ->add_option("-l,--min-length", buildConfig.min_length,
                   normalize_auto_min_length_help());
  min_length_option->default_val(0);
  min_length_option->default_str("auto");
  CLI::Validator min_length_validator;
  min_length_validator.description("auto or non-negative integer");
  min_length_validator.operation(validate_min_length_option);
  min_length_option->check(min_length_validator);
  build
      ->add_option("-t,--threads", buildConfig.threads,
                   "Number of threads for building")
      ->default_val(default_threads);
  build
      ->add_option("--load-factor", buildConfig.load_factor,
                   "IMCF filter load factor")
      ->default_val(0.85);
  build
      ->add_option("--presence-unique-deg", buildConfig.presence_unique_deg,
                   "Degree cutoff (<=) treated as unique signature for coverage meta")
      ->default_val(1)
      ->check(CLI::Range(1, 65535));
  build
      ->add_option("--taxonomy-kind", buildConfig.taxonomy_kind,
                   "Taxonomy source identifier (auto|ncbi|gtdb)")
      ->default_val("auto");
  build
      ->add_option("--taxonomy-version", buildConfig.taxonomy_version,
                   "Taxonomy version label, for example ncbi-taxdump-2025-09-15 or gtdb-rs226")
      ->default_val("auto");
  build->add_flag("-q,--quiet", buildQuietRequested, "Quiet output");

  build->callback([&buildConfig, &buildQuietRequested]() {
    validate_build_config(buildConfig, buildQuietRequested);
  });

  // Classify
  // Add --single option
  auto singleOpt = classify
                       ->add_option("-i,--single", classifyConfig.singleFiles,
                                    "Input file for classifying")
                       ->check(CLI::ExistingFile);

  // Add --paired option
  classify
      ->add_option("-p,--paired", classifyConfig.pairedFiles,
                   "Paired input files for classifying")
      ->check(CLI::ExistingFile)
      ->excludes(singleOpt); // Ensure that single and paired are mutually exclusive

  // Custom validation function to ensure that the --paired option must have an
  // even number of files
  classify->callback([&classifyConfig]() {
    if (!classifyConfig.pairedFiles.empty() &&
        classifyConfig.pairedFiles.size() % 2 != 0) {
      throw CLI::ValidationError(
          "--paired option must have an even number of input files");
    }
  });

  classify
      ->add_option("-o,--output", classifyConfig.outputFile,
                   "Output file for classifying")
      ->default_val("ChimeraClassify");
  classify
      ->add_option("-d,--database", classifyConfig.dbFile,
                   "Database file for classifying")
      ->required()
      ->check(CLI::ExistingFile);
  classify
      ->add_option("-s,--shot-threshold", classifyConfig.shotThreshold,
                   "Shot threshold for classifying")
      ->default_val(0.70);
  auto *firstBetaOpt =
      classify
          ->add_option("--first-filter-beta", classifyConfig.firstFilterBeta,
                       "Keep bins >= beta * max count in first filter")
          ->default_val(0.80);
  classify
      ->add_option("--presence-pi", classifyConfig.presence_pi,
                   "Presence prior P(z=1) for the coverage model (0-1)")
      ->check(CLI::Range(1e-6, 0.5))
      ->default_val(1e-3);
  classify
      ->add_option("--presence-tau", classifyConfig.presence_tau,
                   "Log posterior odds threshold for presence (coverage)")
      ->default_val(4.6);
  classify
      ->add_option("--presence-breadth-bits", classifyConfig.presence_breadth_bits,
                   "Sketch bits for presence breadth (power-of-two suggested)")
      ->check(CLI::Range(64u, 1048576u))
      ->default_val(2048);
  classify
      ->add_option("-t,--threads", classifyConfig.threads,
                   "Number of threads for classifying")
      ->default_val(default_threads);
  classify
      ->add_option("-b,--batch-size", classifyConfig.batchSize,
                   "Batch size for classifying")
      ->default_val(400);
  classify->add_option("--em-iter", classifyConfig.emIter, "EM iteration")
      ->default_val(100);
  classify
      ->add_option("--em-prune-ratio", classifyConfig.em_prune_ratio,
                   "Relative sparsity prune ratio to max expected count in EM")
      ->check(CLI::Range(1e-6, 1.0))
      ->default_val(2e-4);
  classify
      ->add_option("--em-conf-power", classifyConfig.em_conf_power,
                   "Confidence exponent for EM M-step weighting (0 disables)")
      ->check(CLI::Range(0.0, 3.0))
      ->default_val(2.0);
  classify
      ->add_option("--hash-sample-max", classifyConfig.hash_sample_max,
                   "Max sampled hashes for coarse candidate selection")
      ->check(CLI::Range(32.0, 4096.0))
      ->default_val(96);
  classify
      ->add_option("--hash-sample-min", classifyConfig.hash_sample_min,
                   "Min sampled hashes for coarse candidate selection")
      ->check(CLI::Range(8.0, 1024.0))
      ->default_val(16);
  classify
      ->add_option("--idf-max", classifyConfig.idf_max,
                   "Upper cap for IDF weight in evaluate_minimizer")
      ->check(CLI::Range(1.0, 50.0))
      ->default_val(8.0);
  classify
      ->add_option("--post-pi-min", classifyConfig.post_pi_min,
                   "Minimum global class weight")
      ->default_val(5e-4);
  classify->add_flag("--read-evidence", classifyConfig.write_read_evidence,
                     "Write ChimeraReadEvidence.tsv for read-resolved audits");
  // TODO: Deprecated post-processing knobs remain fixed to internal defaults.

  if (argc == 1) {
    std::cout << app.help() << std::endl;
    return 0;
  }

  CLI11_PARSE(app, argc, argv);

  if (firstBetaOpt && firstBetaOpt->count() > 0 &&
      classifyConfig.firstFilterBeta > 0.0) {
    classifyConfig.firstFilterBeta_user = true;
  }

  if (show_version) {
    std::cout << "======================================" << std::endl;
    std::cout << "        Chimera - Metagenomic Tool" << std::endl;
    std::cout << "======================================" << std::endl;
    std::cout << "Version      : " << VERSION_INFO << std::endl;
    std::cout << "Build Date   : " << __DATE__ << " " << __TIME__ << std::endl;
#ifdef __clang__
    std::cout << "Compiled with: Clang " << __clang_major__ << '.'
              << __clang_minor__ << '.' << __clang_patchlevel__ << std::endl;
#elif defined(__GNUC__)
    std::cout << "Compiled with: GCC " << __GNUC__ << '.' << __GNUC_MINOR__ << '.'
              << __GNUC_PATCHLEVEL__ << std::endl;
#elif defined(__VERSION__)
    std::cout << "Compiled with: " << __VERSION__ << std::endl;
#else
    std::cout << "Compiled with: Unknown compiler" << std::endl;
#endif
    std::cout << "OS           : " << detect_os() << std::endl;
    std::cout << "======================================" << std::endl;
    std::cout << "Developed by : Qinzhong Tian" << std::endl;
    std::cout << "Team         : MalabZ" << std::endl;
    std::cout << "Homepage     : https://github.com/LoadStar822/Chimera"
              << std::endl;
    std::cout << "======================================" << std::endl;
    return 0;
  }

  if (*build) {
    ChimeraBuild::run(buildConfig);
  } else if (*classify) {
    ChimeraClassify::run(classifyConfig);
  }

  return 0;
}

File: chimera.py
import argparse
import hashlib
import os
import shutil
import subprocess
import sys
import tarfile
import urllib.request
from pathlib import Path


NCBI_TAXDUMP_URL = "https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz"
NCBI_TAXDUMP_MD5_URL = NCBI_TAXDUMP_URL + ".md5"


def default_threads():
    cpu = os.cpu_count()
    if not cpu or cpu <= 0:
        return 1
    return min(cpu, 192)


def get_downloader():
    from src.download import download as downloader
    return downloader


def get_chimera_path():
    env_path = os.environ.get("CHIMERA_BIN")
    if env_path:
        chimera_path = Path(env_path)
        if chimera_path.is_file() and os.access(chimera_path, os.X_OK):
            return str(chimera_path)
        raise FileNotFoundError(
            f"CHIMERA_BIN points to a missing or non-executable file: {env_path}"
        )
    # Use which command to find the chimera program path
    chimera_path = shutil.which("Chimera")
    if chimera_path is None:
        raise FileNotFoundError(
            "Cannot find 'Chimera' executable. Please ensure it is installed and in your PATH."
        )
    return chimera_path


def _infer_profile_taxonomy_paths(database_path: str):
    db = Path(database_path).resolve()
    search_dirs = []
    if db.parent not in search_dirs:
        search_dirs.append(db.parent)
    stem_path = db.with_suffix("")
    if stem_path.is_dir() and stem_path not in search_dirs:
        search_dirs.append(stem_path)

    taxonomy_meta = None
    taxonomy_info = None
    for base in search_dirs:
        if taxonomy_meta is None:
            candidate = base / "taxonomy.meta"
            if candidate.exists():
                taxonomy_meta = str(candidate)
        if taxonomy_info is None:
            candidate = base / "tax.info"
            if candidate.exists():
                taxonomy_info = str(candidate)
    return taxonomy_meta, taxonomy_info


def _is_auto_or_empty(value) -> bool:
    return value is None or str(value).strip() == "" or str(value).strip().lower() == "auto"


def _read_metadata_file(path: Path):
    values = {}
    if not path.exists():
        return values
    with path.open("r", encoding="utf-8") as handle:
        for line in handle:
            line = line.strip()
            if not line or line.startswith("#") or "=" not in line:
                continue
            key, value = line.split("=", 1)
            values[key.strip()] = value.strip()
    return values


def _apply_build_taxonomy_meta_defaults(args) -> None:
    if not getattr(args, "input", None):
        return
    meta_path = Path(args.input).resolve().parent / "taxonomy.meta"
    meta = _read_metadata_file(meta_path)
    if not meta:
        return
    if _is_auto_or_empty(getattr(args, "taxonomy_kind", None)) and meta.get("taxonomy_kind"):
        args.taxonomy_kind = meta["taxonomy_kind"].lower()
    if _is_auto_or_empty(getattr(args, "taxonomy_version", None)) and meta.get("taxonomy_version"):
        args.taxonomy_version = meta["taxonomy_version"]
    if not getattr(args, "taxonomy_dir", None) and meta.get("taxonomy_dir"):
        args.taxonomy_dir = meta["taxonomy_dir"]


def _directory_has_taxdump(path: Path) -> bool:
    return path.is_dir() and (path / "nodes.dmp").is_file()


def _find_local_taxdump_archive(search_root: Path):
    candidates = [
        search_root / "taxdump.tar.gz",
    ]
    if search_root.is_dir():
        for child in search_root.iterdir():
            if child.is_dir():
                candidates.append(child / "taxdump.tar.gz")
    for candidate in candidates:
        if candidate.is_file() and candidate.stat().st_size > 0:
            return candidate
    return None


def _extract_taxdump_archive(archive_path: Path, extract_dir: Path) -> Path:
    if _directory_has_taxdump(extract_dir):
        return extract_dir
    extract_dir.mkdir(parents=True, exist_ok=True)
    with tarfile.open(archive_path, "r:gz") as archive:
        for member in archive.getmembers():
            if not member.isfile():
                continue
            name = Path(member.name).name
            if not name:
                continue
            target = extract_dir / name
            source = archive.extractfile(member)
            if source is None:
                continue
            with source, target.open("wb") as out:
                shutil.copyfileobj(source, out)
    if not _directory_has_taxdump(extract_dir):
        raise RuntimeError(f"Taxdump archive did not contain nodes.dmp: {archive_path}")
    return extract_dir


def _download_url(url: str, destination: Path) -> None:
    destination.parent.mkdir(parents=True, exist_ok=True)
    partial = destination.with_name(destination.name + ".part")
    try:
        with urllib.request.urlopen(url, timeout=120) as response, partial.open("wb") as out:
            shutil.copyfileobj(response, out)
        partial.replace(destination)
    finally:
        if partial.exists():
            partial.unlink()


def _download_ncbi_taxdump_archive(cache_dir: Path) -> Path:
    archive_path = cache_dir / "taxdump.tar.gz"
    md5_path = cache_dir / "taxdump.tar.gz.md5"
    if not archive_path.is_file() or archive_path.stat().st_size == 0:
        print(f"Downloading NCBI taxdump to {archive_path}")
        _download_url(NCBI_TAXDUMP_URL, archive_path)
    if not md5_path.is_file() or md5_path.stat().st_size == 0:
        _download_url(NCBI_TAXDUMP_MD5_URL, md5_path)
    expected = md5_path.read_text(encoding="utf-8").strip().split()[0].lower()
    digest = hashlib.md5()
    with archive_path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    actual = digest.hexdigest().lower()
    if expected != actual:
        archive_path.unlink(missing_ok=True)
        raise RuntimeError(
            f"Downloaded NCBI taxdump md5 mismatch: expected {expected}, got {actual}"
        )
    return archive_path


def _prepare_build_taxonomy_dir(args) -> None:
    if getattr(args, "no_local_resolution", False):
        return
    _apply_build_taxonomy_meta_defaults(args)
    if getattr(args, "taxonomy_dir", None):
        taxonomy_dir = Path(args.taxonomy_dir).resolve()
        if not _directory_has_taxdump(taxonomy_dir):
            raise FileNotFoundError(
                f"--taxonomy-dir must contain nodes.dmp: {taxonomy_dir}"
            )
        args.taxonomy_dir = str(taxonomy_dir)
        return

    kind = str(getattr(args, "taxonomy_kind", "auto") or "auto").strip().lower()
    if kind == "gtdb":
        return

    if not getattr(args, "input", None):
        return
    input_dir = Path(args.input).resolve().parent
    extract_dir = input_dir / "taxdump"
    if _directory_has_taxdump(extract_dir):
        args.taxonomy_dir = str(extract_dir)
        return
    if _directory_has_taxdump(input_dir):
        args.taxonomy_dir = str(input_dir)
        return

    meta = _read_metadata_file(input_dir / "taxonomy.meta")
    archive_value = meta.get("taxonomy_archive")
    archive_path = None
    if archive_value:
        archive_candidate = Path(archive_value).expanduser()
        if not archive_candidate.is_absolute():
            archive_candidate = input_dir / archive_candidate
        if archive_candidate.is_file() and archive_candidate.stat().st_size > 0:
            archive_path = archive_candidate
    if archive_path is None:
        archive_path = _find_local_taxdump_archive(input_dir)
    if archive_path is None:
        archive_path = _download_ncbi_taxdump_archive(input_dir)
    args.taxonomy_dir = str(_extract_taxdump_archive(archive_path, extract_dir))


def min_length_type(value: str):
    lowered = value.lower()
    if lowered == "auto":
        return "auto"
    try:
        parsed = int(value)
    except ValueError as exc:
        raise argparse.ArgumentTypeError("Minimum length must be an integer or 'auto'") from exc
    if parsed < 0:
        raise argparse.ArgumentTypeError("Minimum length must be >= 0 or 'auto'")
    return parsed


def add_build_arguments(parser, require_input: bool) -> None:
    if require_input:
        parser.add_argument(
            "-i", "--input", required=True, help="Input file for building"
        )
    parser.add_argument(
        "-o", "--output", default="ChimeraDB", help="Output file for building"
    )
    parser.add_argument(
        "--strobe-k",
        type=int,
        default=28,
        help="Strobemer k-mer length",
    )
    parser.add_argument(
        "--strobe-order",
        type=int,
        default=2,
        help="Strobemer order (currently only 2 is supported)",
    )
    parser.add_argument(
        "--strobe-w-min",
        type=int,
        default=12,
        help="Strobemer minimum window",
    )
    parser.add_argument(
        "--strobe-w-max",
        type=int,
        default=32,
        help="Strobemer maximum window",
    )
    parser.add_argument(
        "-l",
        "--min-length",
        type=min_length_type,
        default="auto",
        help="Minimum sequence length for building (auto => strobemer minimum span)",
    )
    parser.add_argument(
        "-t",
        "--threads",
        type=int,
        default=default_threads(),
        help="Number of threads for building",
    )
    parser.add_argument(
        "--load-factor", type=float, default=0.85, help="IMCF filter load factor"
    )
    parser.add_argument(
        "--presence-unique-deg",
        type=int,
        default=1,
        help="Degree cutoff (<=) treated as unique signature for coverage meta",
    )
    parser.add_argument(
        "--taxonomy-kind",
        dest="taxonomy_kind",
        default="auto",
        choices=["auto", "ncbi", "gtdb"],
        help="Taxonomy source identifier (auto/ncbi/gtdb)",
    )
    parser.add_argument(
        "--taxonomy-version",
        dest="taxonomy_version",
        default="auto",
        help="Taxonomy version label, for example ncbi-taxdump-2025-09-15 or gtdb-rs226",
    )
    parser.add_argument(
        "--taxonomy-dir",
        dest="taxonomy_dir",
        default=None,
        help="Directory containing NCBI taxdump nodes.dmp; downloaded automatically for NCBI builds when omitted",
    )
    parser.add_argument(
        "--no-local-resolution",
        action="store_true",
        help="Do not build local read resolution (LPC) data",
    )


def append_build_command_args(command, args) -> None:
    command.extend(["-i", args.input])
    command.extend(["-o", args.output])
    command.extend(["--strobe-k", str(args.strobe_k)])
    command.extend(["--strobe-order", str(args.strobe_order)])
    command.extend(["--strobe-w-min", str(args.strobe_w_min)])
    command.extend(["--strobe-w-max", str(args.strobe_w_max)])
    if args.min_length != "auto":
        command.extend(["-l", str(args.min_length)])
    command.extend(["-t", str(args.threads)])
    command.extend(["--load-factor", str(args.load_factor)])
    command.extend(["--presence-unique-deg", str(args.presence_unique_deg)])
    command.extend(["--taxonomy-kind", str(args.taxonomy_kind)])
    command.extend(["--taxonomy-version", str(args.taxonomy_version)])
    if getattr(args, "taxonomy_dir", None):
        command.extend(["--taxonomy-dir", str(args.taxonomy_dir)])
    if getattr(args, "no_local_resolution", False):
        command.append("--no-local-resolution")


def parse_arguments():
    if len(sys.argv) > 1 and sys.argv[1] == "download":
        parser = argparse.ArgumentParser(
            description="Chimera - A versatile tool for metagenomic classification"
        )
        subparsers = parser.add_subparsers(dest="command", required=False)
        subparsers.add_parser(
            "download", help="Download NCBI database sequences and resources"
        )

        args = parser.parse_args(sys.argv[1:2])
        return args

    parser = argparse.ArgumentParser(
        description="Chimera - A versatile tool for metagenomic classification"
    )

    parser.add_argument(
        "-v", "--version", action="store_true", help="Show version information"
    )

    subparsers = parser.add_subparsers(dest="command", required=False)

    # Download subcommand
    subparsers.add_parser(
        "download", help="Download NCBI database sequences and resources"
    )

    # Build subcommand
    build_parser = subparsers.add_parser(
        "build", help="Build a taxonomic sequence database"
    )
    add_build_arguments(build_parser, require_input=True)

    # Download and Build combined subcommand
    download_build_parser = subparsers.add_parser(
        "download_and_build",
        help="Download NCBI database sequences and resources and build a taxonomic sequence database",
    )
    add_build_arguments(download_build_parser, require_input=False)

    # Classify subcommand
    classify_parser = subparsers.add_parser("classify", help="Classify sequences")
    classify_inputs = classify_parser.add_mutually_exclusive_group(required=True)
    classify_inputs.add_argument(
        "-i",
        "--single",
        nargs="+",
        help="Input files for classifying (supports multiple files)",
    )
    classify_inputs.add_argument(
        "-p", "--paired", nargs="+", help="Paired input files for classifying"
    )
    classify_parser.add_argument(
        "-o",
        "--output",
        default="ChimeraOutput",
        help="Output directory for classify, evidence, and profile results",
    )
    classify_parser.add_argument(
        "-d", "--database", required=True, help="Database file for classifying"
    )
    classify_parser.add_argument(
        "-t",
        "--threads",
        type=int,
        default=default_threads(),
        help="Number of threads for classifying",
    )
    classify_parser.add_argument(
        "-b", "--batch-size", type=int, default=400, help="Batch size for classifying"
    )
    classify_parser.add_argument(
        "--no-local-resolution",
        action="store_true",
        help="Disable local read resolution (LPC) at classify time",
    )
    # Auxiliary profile utilities. Native abundance profiles are written by
    # `classify`; this command is for legacy aggregate conversion and Krona.
    profile_parser = subparsers.add_parser(
        "profile",
        help="Auxiliary profile conversion; classify already writes ChimeraProfile.tsv",
    )
    profile_parser.add_argument(
        "-i",
        "--input",
        nargs="+",
        required=True,
        help="Existing ChimeraEvidence.tsv or aggregate taxid/count table",
    )
    profile_parser.add_argument(
        "-o", "--output", default="ChimeraProfile", help="Output prefix"
    )
    profile_parser.add_argument(
        "-k", "--krona", action="store_true", help="Generate Krona chart"
    )
    profile_parser.add_argument(
        "--taxonomy-kind",
        dest="taxonomy_kind",
        default="auto",
        choices=["auto", "ncbi", "gtdb"],
        help="Taxonomy source for auxiliary profile conversion",
    )
    profile_parser.add_argument(
        "--taxonomy-info",
        dest="taxonomy_info",
        default=None,
        help="Path to the tax.info file produced during build, used by GTDB and other custom taxonomies",
    )
    profile_parser.add_argument(
        "--taxonomy-meta",
        dest="taxonomy_meta",
        default=None,
        help="Path to taxonomy.meta metadata; if omitted, the input directory will be probed",
    )
    if len(sys.argv) == 1:
        parser.print_help(sys.stderr)
        sys.exit(1)

    args = parser.parse_args()

    return args


def run_chimera(args, chimera_path=None):
    if args.command == "download":
        downloader = get_downloader()
        cmd_args = sys.argv
        try:
            download_index = cmd_args.index("download")
            download_args = cmd_args[download_index + 1 :]

            if download_args:
                downloader.download(interactive=False, raw_args=download_args)
            else:
                downloader.download(interactive=True)
        except ValueError:
            downloader.download(interactive=True)
        return 0

    if args.command == "download_and_build":
        downloader = get_downloader()
        options = downloader.download(interactive=True)
        args.command = "build"
        args.input = os.path.join(options.output_dir, "target.tsv")
        resolved_kind = getattr(options, "taxonomy_kind_resolved", None)
        if resolved_kind is None:
            resolved_kind = getattr(options, "taxonomy_mode", None)
        if resolved_kind is None:
            resolved_kind = "auto"
        args.taxonomy_kind = resolved_kind
        resolved_version = getattr(options, "taxonomy_version_resolved", None)
        if resolved_version is None:
            resolved_version = getattr(options, "version_label", None)
        if not resolved_version:
            resolved_version = "auto"
        args.taxonomy_version = resolved_version

    if args.command == "profile":
        from src.profile import profile

        if args.taxonomy_meta is None and args.input:
            candidate_meta = Path(args.input[0]).resolve().parent / "taxonomy.meta"
            if candidate_meta.exists():
                args.taxonomy_meta = str(candidate_meta)
        if args.taxonomy_info is None:
            if args.taxonomy_meta:
                meta_dir = Path(args.taxonomy_meta).resolve().parent
                candidate_info = meta_dir / "tax.info"
                if candidate_info.exists():
                    args.taxonomy_info = str(candidate_info)
            if args.taxonomy_info is None and args.input:
                candidate_info = Path(args.input[0]).resolve().parent / "tax.info"
                if candidate_info.exists():
                    args.taxonomy_info = str(candidate_info)
        if isinstance(args.output, str) and args.output.lower().endswith(".tsv"):
            args.output = args.output[: -len(".tsv")]
        if args.krona:
            downloader = get_downloader()
            from src.profile import conversion2Krona

            print("Converting file to Krona format...")
            conversion2Krona.convert_multiple_files_to_krona_format(
                args.input,
                args.output,
                taxonomy_kind=args.taxonomy_kind,
                taxonomy_info=args.taxonomy_info,
                taxonomy_meta=args.taxonomy_meta,
            )
            print("Conversion completed.")
            print("Generating Krona chart...")
            downloader.run_command(
                ["ktImportText", args.output + ".tsv", "-o", args.output + ".html"]
            )
            print("Krona chart generated.")
        try:
            profile.process_file(
                args.input,
                args.output,
                taxonomy_kind=args.taxonomy_kind,
                taxonomy_info=args.taxonomy_info,
                taxonomy_meta=args.taxonomy_meta,
            )
        except ValueError as exc:
            print(f"Profile failed: {exc}")
            return 1
        return 0

    if chimera_path is None:
        chimera_path = get_chimera_path()
    command = [chimera_path]

    if args.version:
        command.append("--version")

    if args.command:
        command.append(args.command)

    if args.command == "build":
        _prepare_build_taxonomy_dir(args)
        append_build_command_args(command, args)

    elif args.command == "classify":
        output_dir = Path(args.output)
        output_dir.mkdir(parents=True, exist_ok=True)
        classify_output = output_dir / "ChimeraClassify.tsv"
        profile_output = output_dir / "ChimeraProfile.tsv"
        if args.single:
            command.extend(["-i"] + args.single)
        if args.paired:
            command.extend(["-p"] + args.paired)
        command.extend(["-o", str(classify_output)])
        command.extend(["-d", args.database])
        command.extend(["-t", str(args.threads)])
        command.extend(["-b", str(args.batch_size)])
        if getattr(args, "no_local_resolution", False):
            command.append("--no-local-resolution")
    if args.command == "classify":
        result = subprocess.run(command)
        if result.returncode != 0:
            return result.returncode
        if not profile_output.is_file():
            print(f"Classify did not produce native profile: {profile_output}")
            return 1
        return 0
    downloader = get_downloader()
    downloader.run_command(command)
    return 0


def main():
    args = parse_arguments()
    try:
        raise SystemExit(run_chimera(args) or 0)
    except (FileNotFoundError, RuntimeError) as exc:
        print(f"Error: {exc}", file=sys.stderr)
        raise SystemExit(1)


if __name__ == "__main__":
    main()
